docs/index.md

```diff
@@ -58,3 +58,60 @@ another route, based on the changes that the filters made to the
 request.
 
 ![Skipper's request and response processing ](/img/req-and-resp-processing.svg)
+
+#### Routing mechanism
+
+The routing executes the following steps in the typical case:
+
+1. Select the best fitting route by matching the request against the
+   predicates. When no route found, respond with the default status code
+   (the default of the default is 404).
+
+2. Execute the filters defined in the route in normal order on the
+   request. The filters may or may not alter the request.
+
+3. Forward the request to the backend defined by the route and receive
+   a response.
+
+4. Execute the filters defined in the route in reverse order on the
+   response. The filters may or may not alter the response.
+
+5. Respond to the incoming request with the resulting response.
+
+#### Route matching
+
+Skipper can handle a relatively large number of routes acceptable
+performance, while being able to use any attribute of the incoming HTTP
+requests to distinguis between them. In order to be able to do so, the
+path matching predicates (Path() and PathSubtree() but not PathRegexp())
+have a special role during route matching, which is a tradeoff by
+design, and needs to be kept in mind to understand in some cases why a
+certain route was matched for a request instead of another.
+
+The route matching logic can be summed up as follows:
+
+1. Lookup in the path tree based on the Path() and the PathSubtree()
+   predicates, using the path component of the incoming request's URI. Then
+   the remaining predicates of the found route(s) are evaluated.
+
+    * the path lookup is a radix tree with O(log(n)) time complexity
+
+    * in case of intersecting paths, the more specific path is matched in
+     the tree
+
+    * PathRegexp() is not used in the tree, but it is evaluated only after
+     Path() or PathSubtree(), just like e.g. Method() or Host().
+
+2. _If_ step #1 matches multiple routes, which means there are multiple
+   routes in the same position of the path tree, and all other predicates
+   match the request, too, then the route with the most defined predicates
+   is matched.
+
+    * this is an O(n) lookup, but only on the same leaf
+
+    * the root of the tree is considered a single leaf, so if not using the
+      Path() or PathSubtree() predicates, the entire lookup will become O(n)
+      over all the routes.
+
+3. _If_ #2 results in multiple matching routes, then one route will be
```

Comment:
nitpicking: What about *then one route will be selected non-deterministically*?

Reply:
that would actually change the meaning. Stating unspecified leaves it more open for future changes without boundaries.